force-app/main/default/classes/DMLOperations.cls

```diff
@@ -144,9 +197,44 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
+        // Query the Account based on the provided accountName
+        List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName LIMIT 1];
+        Account acc;
+        // Check if any account was found
+        if (accounts.isEmpty()) {
+            acc = new Account(Name = accountName);
+            insert acc;
+        } else {
+            // Retrieve the first Account
+            acc = accounts[0];
+        }
+        // Set to store unique Opportunity names
+        Set<String> existingOpportunityNames = new Set<String>();
+        // Retrieve existing Opportunity names for the Account
+        List<Opportunity> existingOpps = [SELECT Name FROM Opportunity WHERE AccountId = :acc.Id];
+        for (Opportunity existingOpp : existingOpps) {
+            existingOpportunityNames.add(existingOpp.Name);
+        }
+        // List to hold Opportunity records
+        List<Opportunity> opportunitiesToUpsert = new List<Opportunity>();
         // Loop through each Opportunity name
+        for (String oppName : oppNames) {
+            // Check if the name already exists
+            if (existingOpportunityNames.contains(oppName)) {
+                continue;
+            }
```

Comment:
This works pretty well, but you can cut down on some headache by using maps. I think this is a clever solution, though.